res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,921 @@
+        case Prime4.padMode.HOTCUE:
+            mode = modes.hotcue;
+            break;
+        case Prime4.padMode.LOOP:
+            mode = modes.loop;
+            /*
+            // TODO: Implement autoloop mode
+            if (Prime4.PadSection.currentMode === modes.loop) {
+                mode = modes.autoloop;
+            } else {
+                mode = modes.loop;
+            }
+            */
+            break;
+        case Prime4.padMode.ROLL:
+            mode = modes.roll;
+            break;
+        case Prime4.padMode.SLICER:
+            mode = modes.sampler;
+            break;
+        }
+
+        return mode;
+    };
+
+    this.offset = offset;
+
+    this.padModeSelectLeds = new components.Component({
+        trigger: function() {
+            /*
+            modes.forEachComponent(mode => {
+                if (!(mode instanceof components.ComponentContainer)) {
+                    //throw "padmode is not a container as expected";
+                    console.log("padmode is not a container as expected");
+                }
+                //midi.sendShortMsg(0x94 + offset, mode.ledControl, theContainer.currentMode === mode ? mode.colourOn : mode.colourOff);
+                console.log(mode, mode.ledControl, mode.colourOn, mode.colourOff);
+            });
+            */
+            for (const mode of Object.values(modes)) {
+                console.log(mode, mode.ledControl, mode.colourOn, mode.colourOff);
+                midi.sendShortMsg(0x94 + offset, mode.ledControl, theContainer.currentMode === mode ? mode.colourOn : mode.colourOff);
+            }
+        },
+    }, false);
+
+    // Function for switching between pad modes
+    this.setPadMode = function(control) {
+        const newMode = controlToPadMode(control);
+
+        // Exit early if requested mode is already active or unavailable
+        if (newMode === this.currentMode || newMode === undefined) {
+            return;
+        }
+
+        // Disable LED of current mode button
+        if (this.currentMode) {
+            // Disconnect pads from current mode
+            this.currentMode.forEachComponent(function(component) {
+                component.disconnect();
+                component.outConnect = false;
+            });
+        }
+
+        // Connect pads to new mode
+        newMode.forEachComponent(function(component) {
+            component.outConnect = true;
+            component.connect();
+            component.trigger();
+        });
+
+        this.currentMode = newMode;
+
+        theContainer.padModeSelectLeds.trigger();
+    };
+
+    // Start in Hotcue mode
+    this.setPadMode(Prime4.padMode.HOTCUE);
+
+};
+
+Prime4.PadSection.prototype = Object.create(components.ComponentContainer.prototype);
+
+// Assign mode select buttons in XML file
+Prime4.PadSection.prototype.padModeButtonPressed = function(channel, control, value, _status, _group) {
+    if (value) {
+        this.setPadMode(control);
+    }
+};
```

Comment:
I don't see a good reason to make define it like this. Can you shift it to the padsection constructor?